docs/rfcs/new-storage-architecture.md

```diff
@@ -0,0 +1,218 @@
+# Storage challenges at API layer
+
+API layer relies on MySQL for storage. Ideally, each layer is loosely coupled, and dependencies between each layer must be resolved.
+
+The issues are as follows:
+
+Issue1：RunTransaction, Begin, etc. are called and transaction processing is performed at the API layer.
+Issue2：Assembling the Where clause and Order clause is done at the API layer.
+Issue3：The types of DB instances held by each storage class are not uniform.
+
+I propose the following solutions to the above issues.
+
+## Solution for issue 1
+RunTransaction, Begin, etc. are called and transaction processing is performed at the API layer.
+
+### Example of code that is an issue
+- `RunTransaction`
+    - https://github.com/bucketeer-io/bucketeer/blob/main/pkg/account/api/account.go#L64
+- `Begin`
+    - https://github.com/bucketeer-io/bucketeer/blob/main/pkg/autoops/api/api.go#L185
+
+### Solution
+- Define the necessary API for each storage layer using Interface and abstract the storage layer.
+    - In the storage layer, embed the interface for each storage and move the implementation necessary for DB operations to the storage layer.
+    - The API layer calls the above Interface and does not directly perform DB operations.
+````
+// Define the API required for storage with an interface
+type AccountStorage interface {
+	CreateAccount(ctx context.Context, a *domain.AccountV2) error
+	UpdateAccount(ctx context.Context, a *domain.AccountV2) error
+	DeleteAccount(ctx context.Context, a *domain.AccountV2) error
+	GetAccount(ctx context.Context, email, organizationID string) (*domain.AccountV2, error)
+  ...
+}
+
+// The API service maintains instances based on the required storage interface.
+type AccountService struct {
+	accountStorage    v2.AccountStorage     // storage instance
+	publisher         publisher.Publisher
+	opts              *options
+	logger            *zap.Logger
+}
+
+func (s *AccountService) updateAccount(
+	ctx context.Context,
+	editor *eventproto.Editor,
+	email, organizationID string,
+) error {
+  // No DB operations such as transaction processing are performed, just call Update.
```

Comment:
I have the same question with @hvn2k1, some API will need to operate on more than one table, can you please provide more examples about how to do this transaction in API layer.

Reply:
Thank you everyone for your comments.

I'll certainly think of a way to do this and update the documentation, as I need to consider this case.